### DISCUSSION TITLE
Multiple form fields in one Uploady instance

### DISCUSSION THREAD

**Discussion by theilm** [rpldy/react-uploady#300] [Q&A] [answered]
Hello,
I have a question regarding multiple form fields: Is it possible to have one single Uploady instance with multiple different form fields? Let's say we have one field for a photo and one field for a document in a form. I want to be able to differentiate between both fields (e.g. for preview and validation but also for server side processing). But I would like to have a single queue. Is this possible right now or do I have to use separate Uploady instances for each field?
Thanks!
Thomas

**Comment by yoavniran** [comment]
Hey @theilm 

I think the short answer is: YES, its definitely possible. Check out the [Different Configuration](https://github.com/rpldy/react-uploady/blob/master/guides/DifferentConfiguration.md) guide. It shows how you can have different upload buttons with separate configuration wrapped by a single Uploady.

This is of course accurate for the other input methods (Drag&Drop, Paste, and custom).

Files added by these different inputs are processed by the same Uploady (& Uploader) instance, yet can have different configurations.

I would like to know what you mean by "single queue" though, as it might affect the answer here.

**Comment by theilm** [comment]
Hey @yoavniran,

thanks for your fast response! :)

Regarding the "single queue": I have one button that triggers photo uploads and one button that triggers document uploads. I want to have only one upload at a time, even if I add one batch via photo upload and another batch via document upload button. To my understanding, that works out of the box, if I use only one Uploady/Uploader instance.

I had a look at the different configuration options per field. It's great that you can configure e.g. different endpoints per field. Still, I would need to know which field was used in the hooks (e.g. the useBatchAddListener). I am probably not seeing something obvious... But right now I see no way to know which button / upload field was used inside the listeners. How could I display the file names of the uploaded files  or previews next to the respective field and not globally for all fields?

Thanks again - and thanks for the great package!
Thomas

**Reply by yoavniran** [reply]
This is an interesting use-case I havent considered.
I will look into it, to see if I can come up with an example that could work

**Comment by yoavniran** [comment, marked as answer]
Hey @theilm 
There is a relatively simple way to achieve this with an upload param.
Check out this [code sandbox](https://codesandbox.io/s/react-uploady-two-fields-with-single-uploady-e059m?file=/src/App.js)



https://user-images.githubusercontent.com/1102278/150147888-76618ad1-e689-4f2a-bed7-ed26c87164cf.mp4



It generates an "upload field" component with an upload button and preview that is only showing previews for the respective type.

The key in this case is the ability to customize the [preview items provider method](https://github.com/rpldy/react-uploady/tree/master/packages/ui/upload-preview#custom-batch-items-method)

Even if you dont use Uploady's preview component, this should provide an example on how to achieve such separation in your own event handling.

And of course, the differentiation can be based on any of the upload options you used, as long as they are different between your fields. 

Let me know if this helps and if you have more questions.

**Reply by theilm** [reply]
Hey @yoavniran !

Thanks a lot for the example! That's exactly what I was looking for. 

I had not seen that you have access to the options in the hooks. I was relying on my IDE which was relying on typescript. But apparently there is an issue with the types (or with my IDE ;)): The BatchStartHook has an "items" argument, but it does not exist. And the CreateOptions are missing in the BatchCancellableEventHook (I use the useBatchAddListener to show the preview even before the batch has started). Perhaps you could have a look at the types if you find the time. 

But it is working now for me exactly as I hoped, so thanks again for your help!

**Reply by yoavniran** [reply]
Great to hear!

The incorrect batch-start type is something I fixed yesterday (will release shortly). It indeed does **not** receive `items`. Only: `(batch, options)`

The finish/cancel/finalize events don't currently receive the options object. I will look into adding it.

However, you can easily tie things with cancel/finalize events by storing the id of the batch your field is listening to and then use it in other hooks. Using the same custom hook provided to the Preview component.

I updated the sandbox to show this. Of course, the code there is very naive and doesnt take into account edge-cases, so I recommend using it only as a starting point.

**Reply by theilm** [reply]
Hey @yoavniran ,

a little follow up question to this topic: Is it possible to have different input-field options for different upload fields? Let's say, in your codebox example the images should only accept image mime types and should only accept 1 file, the documents should only accept pdfs and multiple files.

As far as I have seen, these options cannot be overwritten like the destination or other upload options. Right?

Thanks!
Thomas

**Reply by yoavniran** [reply]
hi @theilm Uploady maintains a single input-field only so controlling the "accept" property can only be done once.
You are correct in saying it cannot be overridden like most other props

**Reply by theilm** [reply]
Thanks for your reply! :+1:
